app/main.py

```diff
@@ -1,214 +1,40 @@
-            "power": 50,
-        },
-        "potion": None,
-    },
-    "arthur": {
-        "name": "Artur",
-        "power": 45,
-        "hp": 75,
-        "armour": [
-            {
-                "part": "helmet",
-                "protection": 15,
-            },
-            {
-                "part": "breastplate",
-                "protection": 20,
-            },
-            {
-                "part": "boots",
-                "protection": 10,
-            }
-        ],
-        "weapon": {
-            "name": "Two-handed Sword",
-            "power": 55,
-        },
-        "potion": None,
-    },
-    "mordred": {
-        "name": "Mordred",
-        "power": 30,
-        "hp": 90,
-        "armour": [
-            {
-                "part": "breastplate",
-                "protection": 15,
-            },
-            {
-                "part": "boots",
-                "protection": 10,
-            }
-        ],
-        "weapon": {
-            "name": "Poisoned Sword",
-            "power": 60,
-        },
-        "potion": {
-            "name": "Berserk",
-            "effect": {
-                "power": +15,
-                "hp": -5,
-                "protection": +10,
-            }
-        }
-    },
-    "red_knight": {
-        "name": "Red Knight",
-        "power": 40,
-        "hp": 70,
-        "armour": [
-            {
-                "part": "breastplate",
-                "protection": 25,
-            }
-        ],
-        "weapon": {
-            "name": "Sword",
-            "power": 45
-        },
-        "potion": {
-            "name": "Blessing",
-            "effect": {
-                "hp": +10,
-                "power": +5,
-            }
-        }
-    }
-}
+from app.items import Weapon, Armour, Potion
+from app.warriors import Knight
+from app.data import KNIGHTS
 
 
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
+def battle(knights_config: dict) -> dict:
+    knights = {}
 
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
+    for warrior, config in knights_config.items():
+        knight = Knight(name=config["name"],
+                        power=config["power"],
+                        hp=config["hp"])
 
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
+        knight.weapon_up(Weapon(name=config["weapon"]["name"],
+                                power=config["weapon"]["power"]))
 
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
+        for armour in config["armour"]:
+            knight.armour_up(Armour(part=armour["part"],
+                                    protection=armour["protection"]))
 
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
+        if config["potion"]:
```

Comment:
It seems better to transfer all the logic of working with the "potion" to the "use_potion" method of the "Knight" class. And in the main file just call them.